### PR TITLE
option for a seperate audit log file

### PR DESCRIPTION
**Pull request by blizzz**
solves #9166 

* [x] log classes are not static anymore
* [x] apps can have their own log file with code reuse
* [x] admin audit makes use of it
* [x] check rotating magic
* [x] logreader reads old, static, private classes :open_mouth:  → https://github.com/nextcloud/logreader/pull/71
* [x] log to seperate file by default
* [x] ~~evaluate gui for changing log location for admin_audit~~ we don't have this for the general log file either, can be controlled via ``config:app:set admin_audit logfile --value="/var/log/nextcloud/audit.log"`` rotation file size follows general setting

### CODE REVIEW

File: lib/private/Log/Syslog.php [hunk 1]
```diff
@@ -25,21 +25,20 @@
 
 namespace OC\Log;
 
-class Syslog {
-	static protected $levels = array(
+use OCP\IConfig;
+use OCP\Log\IWriter;
+
+class Syslog implements IWriter {
+	static protected $levels = [
```
Comment (icewind1991) [inline lib/private/Log/Syslog.php]: you can use class constant arrays with the currently supported php versions for a bit cleaner code

File: lib/private/Log/Syslog.php [hunk 2]
```diff
@@ -25,21 +25,20 @@
...
 
-	/**
-	 * Init class data
-	 */
-	public static function init() {
-		openlog(\OC::$server->getSystemConfig()->getValue("syslog_tag", "ownCloud"), LOG_PID | LOG_CONS, LOG_USER);
-		// Close at shutdown
+	public function __construct(IConfig $config) {
+		openlog($config->getSystemValue('syslog_tag', 'ownCloud'), LOG_PID | LOG_CONS, LOG_USER);
 		register_shutdown_function('closelog');
```
Comment (icewind1991) [inline lib/private/Log/Syslog.php]: do this with a destructor now?

File: lib/private/Log.php [hunk 3]
```diff
@@ -70,27 +72,19 @@ class Log implements ILogger {
...
+	 * @param IConfig $config the system config object
 	 * @param Normalizer|null $normalizer
 	 * @param IRegistry|null $registry
 	 */
-	public function __construct($logger = null, SystemConfig $config = null, $normalizer = null, IRegistry $registry = null) {
+	public function __construct(IWriter $logger, IConfig $config = null, $normalizer = null, IRegistry $registry = null) {
 		// FIXME: Add this for backwards compatibility, should be fixed at some point probably
 		if ($config === null) {
-			$config = \OC::$server->getSystemConfig();
+			$config = \OC::$server->getConfig();
```
[the remark this answers is not in the document]
Comment (icewind1991) [reply lib/private/Log.php]: Fairly sure switching from systemconfig to config will break yes

**Review by icewind1991 — DISMISSED** [review]

File: lib/private/Log.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (blizzz) [reply lib/private/Log.php]: gnah :(

File: lib/private/Log/Syslog.php (hunk 1 above)
Comment (blizzz) [reply lib/private/Log/Syslog.php]: missed that one, thanks for the hint

**Comment by blizzz** [comment]
✓ reverted change to IConfig in Log
✓ moved Log constants to ILogger and depracted them in Util
✓ polished Syslog a bit

**Comment by MorrisJobke** [comment]
> ✓ moved Log constants to ILogger and depracted them in Util

Moved to a separate PR: #9308

**Comment by MorrisJobke** [comment]
> Moved to a separate PR: #9308

Let me rebase and fix the conflicts in this PR as the other one was merged.

File: lib/private/Log/Errorlog.php [hunk 4]
```diff
@@ -25,22 +25,17 @@
...
+class Errorlog implements IWriter {
 
 	/**
 	 * write a message in the log
 	 * @param string $app
 	 * @param string $message
 	 * @param int $level
 	 */
-	public static function write($app, $message, $level) {
+	public function write(string $app, $message, int $level) {
```
Comment (rullzer) [inline lib/private/Log/Errorlog.php]: `string $message`?

File: lib/private/Server.php [hunk 5]
```diff
@@ -546,15 +548,19 @@ public function __construct($webRoot, \OC\Config $config) {
...
 
-			return new Log($logger, $config, null, $registry);
+			return new Log($logger, $this->getSystemConfig(), null, $registry);
 		});
 		$this->registerAlias('Logger', \OCP\ILogger::class);
 
+		$this->registerService(ILogFactory::class, function (Server $c) {
+			return new LogFactory($c);
+		});
+		$this->registerAlias('LogFactory', ILogFactory::class);
```
Comment (rullzer) [inline lib/private/Server.php]: Not needed

File: lib/private/Server.php [hunk 6]
```diff
@@ -1528,6 +1534,14 @@ public function getLogger() {
 		return $this->query('Logger');
 	}
 
+	/**
+	 * @return ILogFactory
+	 * @throws \OCP\AppFramework\QueryException
+	 */
+	public function getLogFactory() {
+		return $this->query('LogFactory');
```
Comment (rullzer) [inline lib/private/Server.php]: You can just query for ILogFactory::class

Or you remove this function in general and just query for the class where you need it

File: lib/private/Log/File.php [hunk 7]
```diff
@@ -45,30 +48,26 @@
...
-		 * and can not be created.
-		 */
-		if (!file_exists(self::$logFile)) {
-			if(!is_writable(dirname(self::$logFile))) {
-				self::$logFile = $defaultLogFile;
-			} else {
-				if(!touch(self::$logFile)) {
-					self::$logFile = $defaultLogFile;
-				}
+	public function __construct(string $path, string $fallbackPath = '', IConfig $config) {
```
Comment (rullzer) [inline lib/private/Log/File.php]: Any reason to require IConfig over SystemConfig here? It is internal. And the systemconfig doesn't need DB stuff etc. So less change of cyclic dependencies.

**Review by rullzer — CHANGES_REQUESTED** [review]

File: lib/private/Log/Errorlog.php (hunk 4 above)
Comment (blizzz) [reply lib/private/Log/Errorlog.php]: Log takes arrays when logging exceptions :(

Comment (rullzer) [reply lib/private/Log/Errorlog.php]: fair enough

File: lib/private/Log/File.php (hunk 7 above)
Comment (blizzz) [reply lib/private/Log/File.php]: Probably i had in mind back then making to adding API for it, but went another path… anyway, using SystemConfig makes sense. I'll adapt.

**Review by rullzer — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by eniac111** [comment]
Hey, do you have any documentation on this?

**Comment by blizzz** [comment]
@eniac111 in the description ↑ :speak_no_evil: 

 ``config:app:set admin_audit logfile --value="/var/log/nextcloud/audit.log"``

**Comment by MorrisJobke** [comment]
> @eniac111 in the description ↑ 🙊

Could you add it to the documentation/manual?

**Comment by j-ed** [comment]
@MorrisJobke I just stumbled over the missing documentation of the audit log function too. To be able to improve the documentation I would need some more information about this function, which is enabled by activating the **admin_audit** app.
I wonder if it is correct that only audit information (logins/logoffs) is being logged or if other applications should use it too (I also saw **"trashbin:expire"** log entries in this file)?

**Comment by MorrisJobke** [comment]
> @MorrisJobke I just stumbled over the missing documentation of the audit log function too. To be able to improve the documentation I would need some more information about this function, which is enabled by activating the **admin_audit** app.

This separate log file is already used, but with this there is a way to customize the path of this file via:

```
config:app:set admin_audit logfile --value="/var/log/nextcloud/audit.log"
```



> I wonder if it is correct that only audit information (logins/logoffs) is being logged or if other applications should use it too (I also saw **"trashbin:expire"** log entries in this file)?

Only admin_audit entries are logged and we should not use settings of apps in other apps. So this then needs to be implemented in the trash bin as well.

**Comment by j-ed** [comment]
@MorrisJobke I've createds the following pull request for the documentation change: https://github.com/nextcloud/documentation/pull/1300

Do I understand you correct that it is an error that **"trashbin:expire"** messages are logged in the audit.log? If yes, I would create a separate issue ticket to get that fixed:
```
{"reqId":"1qxu8bvbtOn53RyaB3S4","level":1,"time":"2019-03-12 15:56:04+01:00","remoteAddr":"","user":"--","app":"admin_audit","method":"","url":"\/nextcloud\/occ","message":"Console command executed: trashbin:expire --quiet","userAgent":"--","version":"15.0.5.3"}
```

**Comment by nickvergessen** [comment]
> Do I understand you correct that it is an error that "trashbin:expire" messages are logged in the audit.log? If yes, I would create a separate issue ticket to get that fixed:

All `occ` calls are logged to the auditing and have to be. Since they allowed to disabled apps (e.g. the logging app) it would be crucial to not log that.

**Pull request merged**